set_version.sh

```diff
@@ -0,0 +1,102 @@
+
+// Dev. Note – Archived Documentation: https://tinyurl.com/ujl5qhx ------------
+
+import Foundation
+
+// Foundation extensions ------------------------------------------------------
+extension URL {
+  /// Static property to retreive the current working directory.
+  static let currentDir: URL = {
+    URL(fileURLWithPath: ProcessInfo.processInfo.environment["PWD"]!)
+  }()
+
+  /// Static property for where the xcconfig is written to.
+  static let xcconfigURL: URL = {
+    URL.currentDir.appendingPathComponent("Configuration/Version.xcconfig")
+  }()
+}
+
+extension ProcessInfo {
+  var versionArgument: String {
+    arguments.last!
+  }
+}
+
+
+// Bundle Context  ------------------------------------------------------------
+protocol BundleContext {
+    static var bundlePath: String { get }
+    static var plistFile:  String { get }
+
+    static func write(version: String, to url: URL) throws
+}
+
+extension BundleContext {
+  static var bundle: Bundle? {
+    Bundle(url: URL.currentDir.appendingPathComponent(Self.bundlePath))
+  }
+}
+
+struct Settings: BundleContext {
+  static let bundlePath: String = "DuckDuckGo/Settings.bundle"
+  static let plistFile:  String = "Root.plist"
+
+  static func write(version: String, to url: URL) throws {
+    let data  = try Data(contentsOf: url)
+    var plist = try PropertyListSerialization.propertyList(from: data, options: [], format: nil) as! NSDictionary
+    plist     = plist.mutableCopy() as! NSMutableDictionary
+
+    guard var specifiers = plist["PreferenceSpecifiers"] as? [[String:Any]] else {
+      throw CocoaError(.fileReadCorruptFile)
+    }
+
+    guard let index = specifiers.firstIndex(where: { ($0["Key"] as? String) ?? "" == "version" }) else {
+      throw CocoaError(.keyValueValidation)
+    }
+
+    var versionInfo = specifiers[index]
+    versionInfo["DefaultValue"] = version
+
+    specifiers[index] = versionInfo
+    plist.setValue(specifiers, forKey: "PreferenceSpecifiers")
+
+    try plist.write(to: url)
+  }
+}
+
+// Property List File  Wrapper ------------------------------------------------
+struct PropertyList<Context: BundleContext> {
+  /**
+   *
+   */
+  init() throws {
+    let components = Context.plistFile.components(separatedBy: ".")
+    guard let plistURL = Context.bundle?.url(forResource: components.first!, withExtension: components.last!) else {
+      throw URLError(.badURL)
+    }
+
+    self._plistURL = plistURL
+  }
+
+  private let _plistURL: URL
+
+  func write(version: String) throws {
+    try Context.write(version: version, to: self._plistURL)
+  }
+}
+
+// ----------------------------------------------------------------------------
+let version = ProcessInfo.processInfo.versionArgument
+
+do {
+  let xcconfigContent = "MARKETING_VERSION = \(version)"
+  try xcconfigContent.write(to: .xcconfigURL, atomically: true, encoding: .utf8)
+
+  let settings = try PropertyList<Settings>()
+  try settings.write(version: version)
+}
+catch {
+  print(error)
+}
+
```

Comment:
This is pretty nice implementation that uses swift, but seems a bit comlplex. Have you considered using Mac OS tools instead?

Example:
```#!/usr/bin/env bash

if [ -z "$1" ]; then
   echo Usage: ./set_version NUMBER
   echo Example: ./set_version 7.100.1
   exit 1
fi

echo "MARKETING_VERSION = $1" > Configuration/Version.xcconfig
/usr/libexec/PlistBuddy -c "Set :PreferenceSpecifiers:0:DefaultValue $1" DuckDuckGo/Settings.bundle/Root.plist
```

Reply:
Ah! _PlistBuddy_ is **so** much simpler—the only caveat being the hard-coded index (which could be easily forgotten about if more things get added in the future).

I ran your implementation and got the same diff output! This is awesome feedback; thanks so much. I'll squash the commits and update the PR.